File: skills/markdown-preview-approved/preview_api.py
#!/usr/bin/env python3

from __future__ import annotations

import argparse
import json
import sys
import urllib.error
import urllib.parse
import urllib.request
from pathlib import Path


DEFAULT_BASE_URL = "http://localhost:3180"


def main() -> int:
    parser = argparse.ArgumentParser(
        description="Call the TalkAnnotate preview service without writing ad-hoc scripts.",
    )
    parser.add_argument(
        "--base-url",
        default=DEFAULT_BASE_URL,
        help=f"Preview service base URL (default: {DEFAULT_BASE_URL})",
    )

    subparsers = parser.add_subparsers(dest="command", required=True)

    subparsers.add_parser("documents")

    document_parser = subparsers.add_parser("document")
    document_parser.add_argument("--id", required=True)
    document_parser.add_argument("--version", type=int)

    versions_parser = subparsers.add_parser("versions")
    versions_parser.add_argument("--id", required=True)

    annotations_parser = subparsers.add_parser("annotations")
    annotations_parser.add_argument("--id", required=True)
    annotations_parser.add_argument("--version", type=int)

    publish_parser = subparsers.add_parser("publish")
    publish_parser.add_argument("--title", required=True)
    publish_parser.add_argument("--id")
    add_text_source_args(publish_parser)

    changelog_parser = subparsers.add_parser("set-changelog")
    changelog_parser.add_argument("--id", required=True)
    changelog_parser.add_argument("--version", type=int, required=True)
    add_text_source_args(changelog_parser)

    args = parser.parse_args()

    try:
        if args.command == "documents":
            payload = request_json(args.base_url, "/api/documents")
        elif args.command == "document":
            payload = request_json(
                args.base_url,
                build_path(f"/api/documents/{args.id}/content", args.version),
            )
        elif args.command == "versions":
            payload = request_json(args.base_url, f"/api/documents/{args.id}/versions")
        elif args.command == "annotations":
            payload = request_json(
                args.base_url,
                build_path(f"/api/documents/{args.id}/annotations", args.version),
            )
        elif args.command == "publish":
            content = read_text_input(args)
            body = {
                "title": args.title,
                "content": content,
            }
            if args.id:
                body["id"] = args.id
            payload = request_json(
                args.base_url,
                "/api/documents",
                method="POST",
                body=body,
            )
        else:
            change_log = read_text_input(args)
            payload = request_json(
                args.base_url,
                f"/api/documents/{args.id}/versions/{args.version}/change-log",
                method="PUT",
                body={"changeLog": change_log},
            )
    except RuntimeError as error:
        print(str(error), file=sys.stderr)
        return 1

    json.dump(payload, sys.stdout, ensure_ascii=False, indent=2)
    sys.stdout.write("\n")
    return 0


def add_text_source_args(parser: argparse.ArgumentParser) -> None:
    group = parser.add_mutually_exclusive_group(required=True)
    group.add_argument("--text")
    group.add_argument("--file", type=Path)


def build_path(pathname: str, version: int | None) -> str:
    if version is None:
        return pathname

    return f"{pathname}?{urllib.parse.urlencode({'version': version})}"


def read_text_input(args: argparse.Namespace) -> str:
    if args.text is not None:
        return args.text

    return args.file.read_text(encoding="utf-8")


def request_json(
    base_url: str,
    path: str,
    *,
    method: str = "GET",
    body: dict[str, object] | None = None,
) -> object:
    url = f"{base_url.rstrip('/')}{path}"
    data = None if body is None else json.dumps(body).encode("utf-8")

    request = urllib.request.Request(
        url,
        data=data,
        headers={"Content-Type": "application/json"},
        method=method,
    )

    try:
        with urllib.request.urlopen(request) as response:
            return json.load(response)
    except urllib.error.HTTPError as error:
        payload = error.read().decode("utf-8", errors="replace")
        message = payload
        try:
            parsed = json.loads(payload)
            if isinstance(parsed, dict) and isinstance(parsed.get("message"), str):
                message = parsed["message"]
        except json.JSONDecodeError:
            pass

        raise RuntimeError(f"Request failed ({error.code}): {message}") from error
    except urllib.error.URLError as error:
        raise RuntimeError(f"Request failed: {error.reason}") from error


if __name__ == "__main__":
    raise SystemExit(main())

File: scripts/preview-api.mjs
#!/usr/bin/env node

import { readFile } from 'node:fs/promises'

const [command, ...rest] = process.argv.slice(2)
const options = parseOptions(rest)
const baseUrl = normalizeBaseUrl(options['base-url'] ?? process.env.TALKANNOTATE_BASE_URL)

if (!command) {
  printUsageAndExit()
}

await run(command, options)

async function run(selectedCommand, selectedOptions) {
  switch (selectedCommand) {
    case 'annotations':
      await printJson(
        requestJson(
          buildUrl(
            `/api/documents/${required(selectedOptions.id, '--id')}/annotations`,
            optionalVersion(selectedOptions.version),
          ),
        ),
      )
      return
    case 'document':
      await printJson(
        requestJson(
          buildUrl(
            `/api/documents/${required(selectedOptions.id, '--id')}/content`,
            optionalVersion(selectedOptions.version),
          ),
        ),
      )
      return
    case 'documents':
      await printJson(requestJson('/api/documents'))
      return
    case 'publish': {
      const title = required(selectedOptions.title, '--title')
      const content = await readTextInput(selectedOptions)
      await printJson(
        requestJson('/api/documents', {
          body: JSON.stringify({
            content,
            ...(selectedOptions.id ? { id: selectedOptions.id } : {}),
            title,
          }),
          method: 'POST',
        }),
      )
      return
    }
    case 'set-changelog': {
      const changeLog = await readTextInput(selectedOptions)
      await printJson(
        requestJson(
          `/api/documents/${required(selectedOptions.id, '--id')}/versions/${required(selectedOptions.version, '--version')}/change-log`,
          {
            body: JSON.stringify({ changeLog }),
            method: 'PUT',
          },
        ),
      )
      return
    }
    case 'versions':
      await printJson(requestJson(`/api/documents/${required(selectedOptions.id, '--id')}/versions`))
      return
    default:
      console.error(`Unknown command: ${selectedCommand}`)
      printUsageAndExit(1)
  }
}

async function requestJson(pathname, init) {
  const response = await fetch(`${baseUrl}${pathname}`, {
    headers: {
      'Content-Type': 'application/json',
      ...(init?.headers ?? {}),
    },
    ...init,
  })

  const payload = await response.json().catch(() => null)

  if (!response.ok) {
    const message =
      payload && typeof payload === 'object' && 'message' in payload ? payload.message : response.statusText
    throw new Error(`Request failed (${response.status}): ${message ?? 'Unknown error'}`)
  }

  return payload
}

function buildUrl(pathname, params) {
  const search = new URLSearchParams()
  for (const [key, value] of Object.entries(params ?? {})) {
    if (value !== undefined) {
      search.set(key, String(value))
    }
  }
  const query = search.toString()
  return query ? `${pathname}?${query}` : pathname
}

function optionalVersion(value) {
  return value ? { version: value } : undefined
}

async function readTextInput(selectedOptions) {
  if (selectedOptions.text) {
    return String(selectedOptions.text)
  }

  if (selectedOptions.file) {
    return readFile(String(selectedOptions.file), 'utf8')
  }

  throw new Error('One of --text or --file is required.')
}

function parseOptions(args) {
  const parsed = {}

  for (let index = 0; index < args.length; index += 1) {
    const token = args[index]
    if (!token?.startsWith('--')) {
      continue
    }

    const key = token.slice(2)
    const next = args[index + 1]
    if (!next || next.startsWith('--')) {
      parsed[key] = 'true'
      continue
    }

    parsed[key] = next
    index += 1
  }

  return parsed
}

function required(value, name) {
  if (!value) {
    throw new Error(`${name} is required.`)
  }

  return String(value)
}

function normalizeBaseUrl(value) {
  return (value || 'http://localhost:3180').replace(/\/+$/, '')
}

async function printJson(promise) {
  const payload = await promise
  console.log(JSON.stringify(payload, null, 2))
}

function printUsageAndExit(code = 0) {
  console.error(`Usage:
  node scripts/preview-api.mjs documents
  node scripts/preview-api.mjs publish --title "<title>" --file /absolute/path/to/doc.md [--id <document-id>]
  node scripts/preview-api.mjs document --id <document-id> [--version <n>]
  node scripts/preview-api.mjs versions --id <document-id>
  node scripts/preview-api.mjs annotations --id <document-id> [--version <n>]
  node scripts/preview-api.mjs set-changelog --id <document-id> --version <n> (--text "<log>" | --file /absolute/path/to/change-log.md)

Optional:
  --base-url <url>   Override service base URL (default: http://localhost:3180)`)
  process.exit(code)
}
